File: Algorithm/07、归并排序/merge.php
<?php


$a = [0, 9, 1, 5, 8, 3, 7, 6, 4, 2];


$d = merge($a);

function merge($arr)
{
    $len = count($arr);
    if ($len < 2) {
        return $arr;
    }
    $middle = floor($len / 2);
    $left   = array_slice($arr, 0, $middle);
    $right  = array_slice($arr, $middle);

    return mergeArr(merge($left), merge($right));

}

function mergeArr($left, $right)
{


    $result = [];

    while (count($left) > 0 && count($right) > 0) {
        if ($left[0] <= $right[0]) {
            $result[] = array_shift($left);
        } else {
            $result[] = array_shift($right);
        }
    }

    while (count($left)) {
        $result[] = array_shift($left);
    }
    while (count($right)) {
        $result[] = array_shift($right);
    }
//    echo implode(',', $result) . PHP_EOL; // 完成一次增量输出一次结果

    return $result;
}

File: Algorithm/07、归并排序/merge.go
package main

import "fmt"

func main() {
	array := []int{5, 9, 8, 7, 6, 4, 3, 2, 1}
	fmt.Println(array)
	array =	mergeSort(array)
	fmt.Println(array)

}

func mergeSort(array []int) []int {
	length := len(array)
	if length < 2 {
		return array
	}
	mid := length / 2
	left := mergeSort(array[:mid])
	right := mergeSort(array[mid:])

	return merge(left, right)
}

func merge(left, right []int) (result []int) {
	i, j := 0, 0
	for i < len(left) && j < len(right) {
		if left[i] < right[j] {
			result = append(result, left[i])
			i++
		} else {
			result = append(result, right[j])
			j++
		}
	}

	result = append(result, left[i:]...)
	result = append(result, right[j:]...)
	return
}
